plugin/src/main/java/com/playmonumenta/scriptedquests/commands/ImprovedClear.java

```diff
@@ -0,0 +1,212 @@
+import io.github.jorelali.commandapi.api.arguments.IntegerArgument;
+import io.github.jorelali.commandapi.api.arguments.StringArgument;
+import io.github.jorelali.commandapi.api.arguments.TextArgument;
+
+public class ImprovedClear {
+	@SuppressWarnings("unchecked")
+	public static void register() {
+		//TODO: Make this work with actual permissions
+		CommandPermission perms = CommandPermission.fromString("scriptedquests.improvedclear");
+		LinkedHashMap<String, Argument> arguments = new LinkedHashMap<>();
+		String[] aliases = {"iclear"};
+
+		arguments.put("target", new EntitySelectorArgument(EntitySelector.MANY_PLAYERS));
+		arguments.put("name", new TextArgument());
+
+		CommandAPI.getInstance().register("improvedclear", perms, aliases, arguments, (sender, args) -> {
+			clearInventory((Collection<Player>)args[0], (String)args[1], -1, true, "");
+		});
+
+		arguments.put("maxAmount", new IntegerArgument());
+
+		CommandAPI.getInstance().register("improvedclear", perms, aliases, arguments, (sender, args) -> {
+			clearInventory((Collection<Player>)args[0], (String)args[1], (Integer)args[2], true, "");
+		});
+
+		arguments.put("clearShulkers", new BooleanArgument());
+
+		CommandAPI.getInstance().register("improvedclear", perms, aliases, arguments, (sender, args) -> {
+			clearInventory((Collection<Player>)args[0], (String)args[1], (Integer)args[2], (boolean)args[3], "");
+		});
+
+		arguments.put("shulkerLore", new StringArgument());
+
+		CommandAPI.getInstance().register("improvedclear", perms, aliases, arguments, (sender, args) -> {
+			clearInventory((Collection<Player>)args[0], (String)args[1], (Integer)args[2], (boolean)args[3], (String)args[4]);
+		});
+	}
+
+
+	private static void clearInventory(Collection<Player> players, String item, int maxAmount, boolean clearShulkers, String shulkerLore) {
+		for (Player player : players) {
+			clearInventory(player, item, maxAmount, clearShulkers, shulkerLore);
+		}
+	}
+
+	private static void clearInventory(Player player, String itemName, int maxAmount, boolean clearShulkers, String shulkerLock) {
+		Inventory sInventory = player.getInventory();
+		HashMap<Integer, ItemStack> items = new HashMap<>();
+		int itemNum = 0;
+
+		ItemStack[] itemsArray = sInventory.getContents();
+		int arrayCount = 0;
+		for (int i = 0; i < itemsArray.length; i++) {
+			if (itemsArray[i] == null) {
+					continue;
+			}
+			ItemMeta meta = itemsArray[i].getItemMeta();
+			if (meta.getLore() == null || meta.getDisplayName() == null) {
+				continue;
+			} else if (meta.getLore().isEmpty() || meta.getDisplayName().isEmpty()) {
+				continue;
+			}
+
+			if (InventoryUtils.testForItemWithName(itemsArray[i], itemName)) {
+				items.put(arrayCount, itemsArray[i]);
+				arrayCount++;
+			}
+		}
+
+		if (items.size() == 0) {
+			if (clearShulkers) {
+				Iterator<Material> iterator = MaterialUtils.shulkerTypes.iterator();
+				iterateOverShulkerTypes(player, iterator, itemName, maxAmount, shulkerLock, 0);
```

Comment:
The structure of this is really ugly. You should not be passing an iterator from one function to another. 

Something like: 
```
for (Material shulkerMaterial : MaterialUtils.shulkerTypes) {
    clearFromShulkerType(...)
}
```
Note that I don't think you should do this either - this structure is way more complicated than it needs to be.